### PR TITLE
Optimizing ValidateDependenciesChecker for JoinNode

### PR DESCRIPTION
**Pull request by Praveen2112**
The planning time of presto is bit high when compared with the previous presto version (Issue : #11024). So we profiled and figured out that `ValidateDependenciesChecker.java` is taking more to validate a IntermediatePlan. This PR optimizes `ValidateDependenciesChecker.java` so that the planning time can be reduced for queries. 

**Benchmark for initial version**
Benchmark (iterativeOptimizerEnabled) (stage) Mode Cnt Score Error Units
BenchmarkPlanner.planQueries true optimized avgt 20 431.245 ± 15.984 ms/op
BenchmarkPlanner.planQueries true created avgt 20 142.711 ± 2.069 ms/op
BenchmarkPlanner.planQueries false optimized avgt 20 437.861 ± 27.142 ms/op
BenchmarkPlanner.planQueries false created avgt 20 142.856 ± 2.397 ms/op

**Benchmark after optimizing the `JoinNode` validation**
Benchmark (iterativeOptimizerEnabled) (stage) Mode Cnt Score Error Units
BenchmarkPlanner.planQueries true optimized avgt 20 313.418 ± 16.652 ms/op
BenchmarkPlanner.planQueries true created avgt 20 14.891 ± 1.257 ms/op
BenchmarkPlanner.planQueries false optimized avgt 20 298.709 ± 11.603 ms/op
BenchmarkPlanner.planQueries false created avgt 20 14.491 ± 0.460 ms/op

### CODE REVIEW

**Comment by Praveen2112** [comment]
@kokosing Can you please review this also ?

**Comment by atris** [comment]
Please see why the Travis build is failing

**Comment by atris** [comment]
Please perform some wordsmithing on the commit message. It is too verbose to get a gist of changes

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 1]
```diff
@@ -426,8 +427,7 @@ public Void visitIndexSource(IndexSourceNode node, Set<Symbol> boundSymbols)
         @Override
         public Void visitTableScan(TableScanNode node, Set<Symbol> boundSymbols)
         {
-            checkArgument(node.getAssignments().keySet().containsAll(node.getOutputSymbols()), "Assignments must contain mappings for output symbols");
-
+            //We don't have to do a check here as TableScanNode has no dependencies.
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: This should go in  separate commit. Also please also put a space between`//` and `W`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 2]
```diff
@@ -345,20 +346,20 @@ public Void visitJoin(JoinNode node, Set<Symbol> boundSymbols)
...
-
-            checkDependencies(allInputs, node.getOutputSymbols(), "Symbol from join output (%s) not in sources (%s)", node.getOutputSymbols(), allInputs);
-
-            List<Integer> rightSymbolIndexes = node.getRight().getOutputSymbols().stream()
-                    .filter(symbol -> node.getOutputSymbols().contains(symbol))
+            Optional<Integer> lastLeftSymbolIndex = node.getLeft().getOutputSymbols().stream()
                     .map(symbol -> node.getOutputSymbols().indexOf(symbol))
-                    .collect(toImmutableList());
+                    .filter(index -> index >= 0)
+                    .reduce(BinaryOperator.maxBy(Integer::compareTo));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: static import

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 3]
```diff
@@ -345,20 +346,20 @@ public Void visitJoin(JoinNode node, Set<Symbol> boundSymbols)
...
 
-            boolean allLeftBeforeRight = node.getLeft().getOutputSymbols().stream()
-                    .filter(symbol -> node.getOutputSymbols().contains(symbol))
+            Optional<Integer> firstRightSymbolIndex = node.getRight().getOutputSymbols().stream()
                     .map(symbol -> node.getOutputSymbols().indexOf(symbol))
-                    .allMatch(leftIndex -> rightSymbolIndexes.stream()
-                            .allMatch(rightIndex -> rightIndex > leftIndex));
-            checkState(allLeftBeforeRight, "Not all left output symbols are before right output symbols");
+                    .filter(index -> index >= 0)
+                    .reduce(BinaryOperator.minBy(Integer::compareTo));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: static import

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 4]
```diff
@@ -345,20 +346,20 @@ public Void visitJoin(JoinNode node, Set<Symbol> boundSymbols)
...
+
+            if (firstRightSymbolIndex.isPresent() && lastLeftSymbolIndex.isPresent()) {
+                boolean allLeftBeforeRight = lastLeftSymbolIndex.get() < firstRightSymbolIndex.get();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: inline

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 5]
```diff
@@ -345,20 +346,20 @@ public Void visitJoin(JoinNode node, Set<Symbol> boundSymbols)
...
+                checkState(allLeftBeforeRight, "Not all left output symbols are before right output symbols");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: once you inline this, please format this as:

```
checkState(
    condition,
    message);
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 6]
```diff
@@ -345,20 +346,20 @@ public Void visitJoin(JoinNode node, Set<Symbol> boundSymbols)
...
+            Optional<Integer> lastLeftSymbolIndex = node.getLeft().getOutputSymbols().stream()
                     .map(symbol -> node.getOutputSymbols().indexOf(symbol))
-                    .collect(toImmutableList());
+                    .filter(index -> index >= 0)
+                    .reduce(BinaryOperator.maxBy(Integer::compareTo));
 
-            boolean allLeftBeforeRight = node.getLeft().getOutputSymbols().stream()
-                    .filter(symbol -> node.getOutputSymbols().contains(symbol))
+            Optional<Integer> firstRightSymbolIndex = node.getRight().getOutputSymbols().stream()
                     .map(symbol -> node.getOutputSymbols().indexOf(symbol))
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: is `indexOf` O(n)? If so then iterating all n symbols and retrieving their indexes is O(n*n).

Maybe you could do something like:
```
for (int i = 0; i < node.getOutputSymbols().size(); i++) {
  Symbol symbol =  node.getOutputSymbols().get(i);
  if (leftSymbols.contains(symbol)) {
     // update max left Symbol
  } 
  ...
```
?

Comment (Praveen2112) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: `leftSymbols.contains(symbol)` also performs `indexOf(symbol) > 0` internally wouldn't that be a issue ?  In the initial version of the code

```
List<Integer> rightSymbolIndexes = node.getRight().getOutputSymbols().stream()
     .filter(symbol -> node.getOutputSymbols().contains(symbol))
     .map(symbol -> node.getOutputSymbols().indexOf(symbol))
     .collect(toImmutableList());
````
might ideally do the same job twice

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: I am sorry that I wasn't precise enough. 

`leftSymbols.contains(symbol)` -  I meant using `HashSet` for `leftSymbols` here.

Comment (Praveen2112) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: Oh yes that can be done

**Comment by Praveen2112** [comment]
@kokosing Updated the code and also ran the `BenchmarkPlanner` for benchmarking.
For the initial version : 
Benchmark                     (iterativeOptimizerEnabled)    (stage)  Mode  Cnt     Score     Error  Units
BenchmarkPlanner.planQueries                         true  optimized  avgt   20  2199.486 ± 316.666  ms/op
BenchmarkPlanner.planQueries                         true    created  avgt   20   150.316 ±  16.532  ms/op
BenchmarkPlanner.planQueries                        false  optimized  avgt   20  1687.298 ± 215.590  ms/op
BenchmarkPlanner.planQueries                        false    created  avgt   20    79.258 ±   6.999  ms/op

After applying the patch:
Benchmark                     (iterativeOptimizerEnabled)    (stage)  Mode  Cnt     Score     Error  Units
BenchmarkPlanner.planQueries                         true  optimized  avgt   20  1258.897 ±  50.254  ms/op
BenchmarkPlanner.planQueries                         true    created  avgt   20    70.772 ±   6.227  ms/op
BenchmarkPlanner.planQueries                        false  optimized  avgt   20   933.292 ± 208.551  ms/op
BenchmarkPlanner.planQueries                        false    created  avgt   20    72.930 ±  11.245  ms/op

**Comment by Praveen2112** [comment]
@kokosing There was a small mistake in the benchmark so removed it. Will update it

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 7]
```diff
@@ -346,19 +348,19 @@ public Void visitJoin(JoinNode node, Set<Symbol> boundSymbols)
...
-                    .collect(toImmutableList());
-
-            boolean allLeftBeforeRight = node.getLeft().getOutputSymbols().stream()
-                    .filter(symbol -> node.getOutputSymbols().contains(symbol))
-                    .map(symbol -> node.getOutputSymbols().indexOf(symbol))
-                    .allMatch(leftIndex -> rightSymbolIndexes.stream()
-                            .allMatch(rightIndex -> rightIndex > leftIndex));
-            checkState(allLeftBeforeRight, "Not all left output symbols are before right output symbols");
+            int leftMaxPosition = Integer.MIN_VALUE;
+            int rightMinPosition = Integer.MAX_VALUE;
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: but what if the `JoinNode#getOuputSymbols` has no symbols from the `rightRelation` ? So we set that to `Integer.MAX_VALUE` in order to make sure that `rightMinPosition > leftMaxPosition`

[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: then you could have `rightMinPosition > leftMaxPosition or rightMinPosition == -1`

Also, lets use `OptionalInt`.

**Comment by Praveen2112** [comment]
@kokosing Handled those minor changes.

**Benchmark for the initial version** 
BenchmarkPlanner.planQueries                         true  optimized  avgt   20  1542.007 ± 124.085  ms/op
BenchmarkPlanner.planQueries                         true    created  avgt   20    77.899 ±   6.070  ms/op
BenchmarkPlanner.planQueries                        false  optimized  avgt   20   952.575 ± 117.280  ms/op
BenchmarkPlanner.planQueries                        false    created  avgt   20    81.141 ±   4.798  ms/op

**Benchmark after the change**
BenchmarkPlanner.planQueries                         true  optimized  avgt   20  1432.079 ± 120.257  ms/op
BenchmarkPlanner.planQueries                         true    created  avgt   20    74.664 ±   4.194  ms/op
BenchmarkPlanner.planQueries                        false  optimized  avgt   20   892.081 ±  89.001  ms/op
BenchmarkPlanner.planQueries                        false    created  avgt   20    79.808 ±   6.703  ms/op


Though not much of a difference is seen the tpch queries, for queries involving wide tables and more number of joins, the time taken for validation reduces from **1400ms** to **70ms**

**Comment by kokosing** [comment]
> Though not much of a difference is seen the tpch queries, for queries involving wide tables and more number of joins, the time taken for validation reduces from 1400ms to 70ms

Can you please provide a benchmark for this case. This case seems to be very important, having such benchmark will be useful for further optimizations.

**Comment by kokosing** [comment]
Regarding the benchmark numbers, now they are in line with my intuition for current benchmark.

**Comment by Praveen2112** [comment]
@kokosing I took a benchmark on tpch query 5 with modified tables (Number of columns in each table is increased by 50x). 

**Benchmark for initial version**  
Benchmark                     (iterativeOptimizerEnabled)    (stage)  Mode  Cnt    Score    Error  Units
BenchmarkPlanner.planQueries                         true  optimized  avgt   20  431.245 ± 15.984  ms/op
BenchmarkPlanner.planQueries                         true    created  avgt   20  142.711 ±  2.069  ms/op
BenchmarkPlanner.planQueries                        false  optimized  avgt   20  437.861 ± 27.142  ms/op
BenchmarkPlanner.planQueries                        false    created  avgt   20  142.856 ±  2.397  ms/op


**Benchmark after applying patch**
Benchmark                     (iterativeOptimizerEnabled)    (stage)  Mode  Cnt    Score    Error  Units
BenchmarkPlanner.planQueries                         true  optimized  avgt   20  313.418 ± 16.652  ms/op
BenchmarkPlanner.planQueries                         true    created  avgt   20   14.891 ±  1.257  ms/op
BenchmarkPlanner.planQueries                        false  optimized  avgt   20  298.709 ± 11.603  ms/op
BenchmarkPlanner.planQueries                        false    created  avgt   20   14.491 ±  0.460  ms/op


From these numbers we can infer that we have optimized the `ValidateDependenciesChecker` for `JoinNode` when we are validating a intermediate plan (437ms is reduced to 15ms)

**Comment by kokosing** [comment]
> (437ms is reduced to 15ms)

How did you calculate that?

For example from the below:
> Benchmark for initial version
Benchmark (iterativeOptimizerEnabled) (stage) Mode Cnt Score Error Units
BenchmarkPlanner.planQueries true optimized avgt 20 431.245 ± 15.984 ms/op
Benchmark after applying patch
Benchmark (iterativeOptimizerEnabled) (stage) Mode Cnt Score Error Units
BenchmarkPlanner.planQueries true optimized avgt 20 313.418 ± 16.652 ms/op

I see that benchmark is faster about 120ms. Still this is nice improvement.

**Comment by Praveen2112** [comment]
Sorry it was 142ms is reduced to 14ms (`CreatedStage`)

**Comment by kokosing** [comment]
Please update the commit message, so it contains the benchmark results and it is ready to go.

**Comment by Praveen2112** [comment]
@kokosing Updated the commit message
